library/alloc/src/sync.rs

```diff
@@ -2112,55 +2158,48 @@ impl<T: Clone, A: Allocator + Clone> Arc<T, A> {
-        // the memory to be deallocated.
-        //
-        // Use Acquire to ensure that we see any writes to `weak` that happen
-        // before release writes (i.e., decrements) to `strong`. Since we hold a
-        // weak count, there's no chance the ArcInner itself could be
-        // deallocated.
-        if this.inner().strong.compare_exchange(1, 0, Acquire, Relaxed).is_err() {
-            // Another strong pointer exists, so we must clone.
-            // Pre-allocate memory to allow writing the cloned value directly.
-            let mut arc = Self::new_uninit_in(this.alloc.clone());
-            unsafe {
-                let data = Arc::get_mut_unchecked(&mut arc);
-                (**this).write_clone_into_raw(data.as_mut_ptr());
-                *this = arc.assume_init();
+        match this.inner().strong.load(Relaxed) {
+            ONE_STRONG => {
+                // This is the only Arc, and no Weak pointers exist.
+                acquire!(self.inner().strong);
             }
-        } else if this.inner().weak.load(Relaxed) != 1 {
-            // Relaxed suffices in the above because this is fundamentally an
-            // optimization: we are always racing with weak pointers being
-            // dropped. Worst case, we end up allocated a new Arc unnecessarily.
-
-            // We removed the last strong ref, but there are additional weak
-            // refs remaining. We'll move the contents to a new Arc, and
-            // invalidate the other weak refs.
-
-            // Note that it is not possible for the read of `weak` to yield
-            // usize::MAX (i.e., locked), since the weak count can only be
-            // locked by a thread with a strong reference.
-
-            // Materialize our own implicit weak pointer, so that it can clean
-            // up the ArcInner as needed.
-            let _weak = Weak { ptr: this.ptr, alloc: this.alloc.clone() };
-
-            // Can just steal the data, all that's left is Weaks
-            let mut arc = Self::new_uninit_in(this.alloc.clone());
-            unsafe {
-                let data = Arc::get_mut_unchecked(&mut arc);
-                data.as_mut_ptr().copy_from_nonoverlapping(&**this, 1);
-                ptr::write(this, arc.assume_init());
+            ONE_STRONG_WITH_WEAK
+                if this
+                    .inner()
+                    .strong
+                    .compare_exchange(ONE_STRONG_WITH_WEAK, STRONG_DROPPED, Acquire, Relaxed)
+                    .is_ok() =>
+            {
+                // This is the only Arc, but other Weak pointers might exist.
+                // Weak pointers can no longer upgrade at this point, though.
+                if this.inner().weak.load(Relaxed) == ONE_WEAK {
+                    // No Weak pointers exist.
+                    this.inner().strong.store(ONE_STRONG_WITH_WEAK, Release); // Restore the strong count for this Arc.
```

Comment:
At this point would it be safe to return to the state `(ONE_STRONG, NO_WEAK)`?

Reply:
Yes, I think so!